blobstore.wit.md

```diff
@@ -0,0 +1,178 @@
+```
+
+// # wasi-cloud blobstore service definition
+//
+// A blobstore is service that can mange containers and objects.
+//
+// ## Object
+// An object is a named sequence of bytes. Objects can be read or written with a stream interface.
+// Once an object is written its contents cannot be modified.
+// Objects can be any size from 1 byte up to the limits imposed by the underlying store.
+// Object names are unique within their container.
+//
+// ## Container
+// A container is a named collection of objects. Container names can be any utf-8 string.
+// Within any instance of the blobstore interface, all containers share the same "flat" namespace.
+
+
+// wasi-cloud Blobstore service definition
+interface "wasi:blob/blobstore" {
+
+  use wasi:blob/types::{ Error, container-metadata, container-name, object-id }
+
+  // creates a new empty container
+  create-container: func(name: container-name) -> result<container,Error>
+
+  // retrieves a container by name
+  get-container: func(name: container-name) -> result<container, Error>
+
+  // deletes a container and all objects within it
+  delete-container: func(name: container-name) -> result<_, Error>
+
+  // returns true if the container exists
+  container-exists: func(name: container-name) -> result<bool, Error>
+
+  // copies (duplicates) an object, to the same or a different container.
+  // returns an error if the target container does not exist.
+  // overwrites destination object if it already existed.
+  copy-object: func(src: object-id, dest: object-id) -> result<_, Error>
+
+  // moves or renames an object, to the same or a different container
+  // returns an error if the destination container does not exist.
+  // overwrites destination object if it already existed.
+  move-object: func(src:object-id, dest: object-id) -> result<_, Error>
+}
+
+// a Container is a collection of objects
+resource container {
+
+  // returns container name
+  name: func() -> result<string, Error>
+
+  // returns container metadata
+  info: func() -> result<container-metadata, Error>
+
+  // begins reading an object
+  read-object: func(name: object-name) -> result<read-stream, Error>
+
+  // creates or replaces an object.
+  write-object: func(name: object-name) -> result<write-stream, Error>
```

Comment:
What is write-stream type?

There is a "wasi:blob/write-stream" interface defined below but I don't think it's valid in WIT to have a function return a interface. Do you intend to define write-stream as a `resource`, which has a `write` function?

Reply:
fixed